src/v/config/configuration.cc

```diff
@@ -353,15 +357,22 @@ configuration::configuration()
       {.needs_restart = needs_restart::no,
        .example = "LogAppendTime",
        .visibility = visibility::user},
-      model::timestamp_type::create_time)
+      model::timestamp_type::create_time,
+      {model::timestamp_type::create_time, model::timestamp_type::append_time})
   , log_compression_type(
       *this,
       "log_compression_type",
       "Default topic compression type",
       {.needs_restart = needs_restart::no,
        .example = "snappy",
        .visibility = visibility::user},
-      model::compression::producer)
+      model::compression::producer,
+      {model::compression::none,
```

Comment:
is comparison done on the numeric value of the enum, or on a string name for the options? it looks like the former. if so, perhaps we should compare against either the `operator<<` stringified version of the name or have the property take an explicit mapping so we don't internally need to track which enums have a sort of config-abi?

Reply:
It's a tradeoff I guess, between the convenience of being able to use an enum directly as the T in a property<>, vs. the explicit-ness of writing out the mappings in the property definition as the stable external interface.

Doing an explicit mapping in the property is a bit more complicated, because we'd have to be generic on both T and the type of the object we're taking from YAML (our input isn't necessarily a string: config values start out as a YAML::Node which just happens to be a string in this case).  :thinking:

Reply:
IIUC this is an issue with the interface, right? It kinda sorta seems like it could be done with a wrapper:

`property<enum_wrapper<model::compression_type>>`

that maintained the enum type but also had a string constructor would work? it is all pretty low-risk, so I agree that the trade off is reasonable with respect to convenience.

independent of outcome of this side discussion PR looks good

Reply:
I think I'd like to merge this as-is, the part I want to get into 22.1 is the admin API changes, so that enumerations are part of the config schema api from day 1.

I'd like to be a bit clearer somehow on which enums have stringized forms that must be stable, I wonder if maybe this is a case where "enum class" just isn't quite sufficient, and we need a slightly fuller-fat enumeration class that includes "give me a list of your possible values" that enum_property can use to avoid replicating the list of values in the definition of the property.

Reply:
Sounds good. Seems like a bigger thing than I had imagined.

FWIW, I think that the configuration layer can map between the two domains: the external configuration name commitments, and then internal choice of enum whose values are opaque. It's more code to maintain per enum property, but it fully decouples internals from external configs.

Resolving